sips/fork_support.md

```diff
@@ -20,56 +20,260 @@ Client implementation should calculate when a new duty starts to decide if it st
-    PrimusTestnet  = DomainType{0x0, 0x0, PrimusChain, 0x0}
-    ShifuTestnet   = DomainType{0x0, 0x0, ShifuChain, 0x0}
-    ShifuV2Testnet = DomainType{0x0, 0x0, ShifuChain, 0x1}
-    JatoTestnet    = DomainType{0x0, 0x0, JatoChain, 0x1}
+	GenesisMainnet = DomainType{0x0, 0x0, byte(MainnetNetworkID), 0x0}
+	PrimusTestnet  = DomainType{0x0, 0x0, byte(PrimusNetworkID), 0x0}
+	ShifuTestnet   = DomainType{0x0, 0x0, byte(ShifuNetworkID), 0x0}
+	ShifuV2Testnet = DomainType{0x0, 0x0, byte(ShifuNetworkID), 0x1}
+	JatoTestnet    = DomainType{0x0, 0x0, byte(JatoNetworkID), 0x0}
+	JatoV2Testnet  = DomainType{0x0, 0x0, byte(JatoV2NetworkID), 0x1}
 )
 
+// ForkData is a simple structure holding fork information for a specific chain (and its fork)
 type ForkData struct {
-    Epoch  phase0.Epoch
-    Domain DomainType
+	// Epoch in which the fork happened
+	Epoch phase0.Epoch
+	// Domain for the new fork
+	Domain DomainType
+}
+
+func (domainType DomainType) GetNetworkID() NetworkID {
+	return NetworkID(domainType[2])
 }
 
-func (domainType DomainType) GetChain() byte {
-    return domainType[2]
+func (networkID NetworkID) GetForksData() []*ForkData {
+	switch networkID {
+	case MainnetNetworkID:
+		return mainnetForks()
+	default:
+		return []*ForkData{
+			{
+				Epoch:  0,
+				Domain: DomainType{0x0, 0x0, byte(networkID), 0x0},
+			},
+		}
+	}
 }
 
-func (domainType DomainType) GetForksData() []*ForkData {
-    switch domainType.GetChain() {
-    case GenesisChain:
-        return genesisForks()
-    default:
-        return []*ForkData{}
-    }
+// mainnetForks returns all forks for the mainnet chain
+func mainnetForks() []*ForkData {
+	return []*ForkData{
+		{
+			Epoch:  0,
+			Domain: GenesisMainnet,
+		},
+	}
 }
 
-func genesisForks() []*ForkData {
-    return []*ForkData{
-        {
-            Epoch:  0,
-            Domain: GenesisMainnet,
-        },
-    }
+func (networkID NetworkID) DefaultFork() *ForkData {
+	return networkID.GetForksData()[0]
 }
 
+// GetCurrentFork returns the ForkData with highest Epoch smaller or equal to "epoch"
+func (networkID NetworkID) ForkAtEpoch(epoch phase0.Epoch) (*ForkData, error) {
+	// Get list of forks
+	forks := networkID.GetForksData()
+
+	// If empty, raise error
+	if len(forks) == 0 {
+		return nil, errors.New("GetCurrentFork: fork list by GetForksData is empty.")
+	}
+
+	var current_fork *ForkData
+	for _, fork := range forks {
+		if fork.Epoch <= epoch {
+			current_fork = fork
+		}
+	}
+	return current_fork, nil
+}
+
+func (f ForkData) GetRoot() ([]byte, error) {
+	byts, err := json.Marshal(f)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal ForkData")
+	}
+	ret := sha256.Sum256(byts)
+	return ret[:], nil
+}
+```
+
+### Structures access to types
+
+- `Share`: It's more correct for the structure to hold the information on the network that it's in, instead of the network and fork number.
+    Thus, _Share_ will contain a `NetworkID` field instead of a `DomainType` field.
```

Comment:
Is that so important? domain type contains network

Reply:
It's not that important.
It's a two-way relation. We can derive NetworkID from DomainType and vice-versa, so in practice Share can have either one.
I thought it would be more correct if the Share was to contain the NetworkID instead of DomainType. But I can roll back this change. What do you think?

Reply:
I would roll back. This is because:

1. I am not sure it is a two way relation. Note that DomainType has 2 TBD bytes. Are you sure they can always be derived.
2. If it can't be derived, if you look at the diagram you made, for "duties" that don't have a qbft Controller we will have no way to derive domain types.

Reply:
I thought that it was a two-way relation because with the NetworkID we can get the forks (GetForksData()) and then get the domain of the current fork.

Reply:
What about the TBD bytes?

Reply:
Sorry, I mean that the Fork has the whole Domain data associated with an epoch, not just the fork number. So we get the whole Domain (all 4 bytes) through the GetForksData